<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [40]:
# текст запроса
query_3_1 = f'''
select count(id)
from public.vacancies
'''

In [41]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_1480\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

In [42]:
connection.close()

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [46]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [47]:
query_3_1 = f'''
select count(id)
from public.employers
'''# текст запроса

In [48]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_1480\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [49]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [50]:
# текст запроса
query_3_1 = f'''
select count(id)
from public.areas
'''

In [51]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_1480\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [56]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [57]:
# текст запроса
query_3_1 = f'''
select count(id)
from public.industries
'''

In [58]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_1480\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0    294

***

# выводы по предварительному анализу данных
В среднем на каждого работодателя приходится примерно по 2 вакансии, на каждую сферу деятельности - порядка 167 вакансий, на регион - около 36 вакансий. Однако средние цифры могут давать не совсем объективную информацию, поэтому требуется более детальный анализ конкретных вакансий и работодателей.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [59]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [61]:
# текст запроса
query_3_1 = f'''
select count(V.id), A.name
from public.vacancies V 
join areas A on V.area_id = A.id 
group by A.name
order by count(V.id) desc
'''


In [62]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_1480\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count             name
0     5333           Москва
1     2851  Санкт-Петербург
2     2112            Минск
3     2006      Новосибирск
4     1892           Алматы
..     ...              ...
764      1       Тарко-Сале
765      1    Новоаннинский
766      1            Бирск
767      1           Сасово
768      1             Анна

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [63]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [64]:
# текст запроса
query_3_1 = f'''
select count(id)
from public.vacancies  
where salary_from is not null or salary_to is not null
'''

In [65]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_1480\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [66]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [67]:
# текст запроса
query_3_1 = f'''
select avg(salary_from) as "Нижняя граница", avg(salary_to) as "Верхняя граница"
from public.vacancies 
'''

In [68]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_1480\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Нижняя граница  Верхняя граница
0    71064.657901    110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [69]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [70]:
# текст запроса
query_3_1 = f'''
select count(id), schedule, employment
from public.vacancies  
group by schedule, employment
order by count(id) desc
'''

In [71]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_1480\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [5]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [79]:
# текст запроса
query_3_1 = f'''
select count(id), experience as "Требуемый опыт работы"
from public.vacancies 
group by experience
order by count(id) 
'''

In [80]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_1480\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count Требуемый опыт работы
0   1337           Более 6 лет
1   7197             Нет опыта
2  14511         От 3 до 6 лет
3  26152    От 1 года до 3 лет

***

# выводы по детальному анализу вакансий
1. Наибольшее количество вакансий, как и ожидалось, предлагается работодателями из крупных городов, таких как Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. 
2. Приблизительно в половине случаев работодатели в описании вакансии не оговаривают размер заработной платы (или верхнюю и нижнюю границы). В тех объявлениях, где все же указаны границы зарплатной вилки, значения средних верхней и нижней границ варьируют в пределах от 71065 руб. до 110537 руб. 
3. Наиболее часто предлагаемым работодателями сочетанием типа рабочего графика и типа трудоустройства является "полный день-полная занятость".
4. Чаще всего встречаются вакансии с требуемым опытом работы от 1 до 3 лет. Интересным является факт того, что количество вакансий, в которых отсутствует требование к наличию какого-либо опыта работы, больше, чем таковое с требованием к опыту работы более 6-ти лет. Из этого следует, что опытные соискатели менее востребованы, чем претенденты без опыта работы. Однако данное обстоятельство может быть связано с тем, что описание требований к опыту работы иногда приводится в тексте описания самой вакансии и при этом не всегда в явном виде (например, может быть требование к наличию портфолио, а это уже, соответственно, подразумевает наличие опыта). Также это может свидетельствовать о том, что работодатели предлагают больше вакансий, не требующих особых навыков. Однако такая работа является менее оплачиваемой, что подтверждает дополнительный анализ, приведенный ниже.

In [75]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [76]:
# текст запроса
query_3_1 = f'''
select  avg(salary_from), avg(salary_to), experience 
from public.vacancies 
group by experience 
order by avg(salary_from), avg(salary_to) 
'''

In [77]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_1480\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


avg            avg          experience
0   34698.743072   56990.497384           Нет опыта
1   64456.363054   97435.916560  От 1 года до 3 лет
2  112544.075112  171974.030400       От 3 до 6 лет
3  150155.510806  197819.223077         Более 6 лет

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [82]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [83]:
# текст запроса
query_3_1 = f'''
select e.name, count(v.id) 
from public.vacancies v 
join public.employers e on v.employer_id = e.id
group by e.name
order by count(v.id) desc
limit 5
'''

In [84]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_1480\2746117666.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [6]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [7]:
# текст запроса
query_3_1 = f'''
select a.name as "Регион", count(e.id) as "Количество работодателей", count(v.id) as "Количество вакансий"
from public.areas a
join public.employers e on a.id = e.area
left join vacancies v on a.id = v.area_id
where v.id is null
group by a.name
order by count(e.id) desc
'''

In [8]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_11192\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Регион  Количество работодателей  \
0                                       Россия                       410   
1                                    Казахстан                       207   
2                           Московская область                        75   
3                           Краснодарский край                        19   
4                                     Беларусь                        18   
..                                         ...                       ...   
181                                  Миллерово                         1   
182  Михайловка (Уфимский район, Башкортостан)                         1   
183                                    Можайск                         1   
184                                Морозовск\n                         1   
185                                     Несвиж                         1   

     Количество вакансий  
0                      0  
1                      0  
2                      0  
3                      0  
4                      0  
..                   ...  
181                    0  
182                    0  
183                    0  
184                    0  
185                    0  

[186 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [9]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [11]:
# текст запроса
query_3_1 = f'''
select  e.name as "Наименование компании", count(distinct v.area_id) as "Количество регионов"
from public.vacancies v
join public.employers e on e.id = v.employer_id
group by  e.name
order by count(distinct v.area_id) desc
'''

In [12]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_11192\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Наименование компании  Количество регионов
0                      Яндекс                  181
1                  Ростелеком                  152
2                  Спецремонт                  116
3      Поляков Денис Иванович                   88
4                   ООО ЕФИН                    71
...                       ...                  ...
14761                  UniSol                    1
14762            UNISTORY LLC                    1
14763                   UNIT6                    1
14764     United Distribution                    1
14765              UNITEL LLC                    1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [4]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [5]:
# текст запроса
query_3_1 = f'''
select count(e.id)
from public.employers e
left join public.employers_industries ei on e.id = ei.employer_id
where ei.employer_id is null
'''


In [6]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_8724\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [12]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [13]:
# текст запроса
query_3_1 = f'''
select e.name as "Название компании"
from public.employers e 
join public.employers_industries ei on e.id = ei.employer_id
group by e.name
having count(ei.industry_id) = 4
order by e.name
limit 3 
'''

In [14]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_8724\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Название компании
0      101 Интернет
1          21vek.by
2              2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [15]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [16]:
# текст запроса
query_3_1 = f'''
select count(e.id)
from public.employers e 
join public.employers_industries ei on e.id = ei.employer_id
join public.industries i on i.id = ei.industry_id
where i.name = 'Разработка программного обеспечения'
'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_8724\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

***

# выводы по анализу работодателей
1. Из 23501 компаний-работодателей лидером по количеству вакансий (1933) и охвату регионов (181) является Яндекс. В целом видно, что основными работодателями являются достаточно крупные компании, которые, вероятно, в связи с развитием постоянно расширяют штат. Также возможно, что данные работодатели предлагают не самые комфортные условия (сдельная оплата, низкая заработная плата, система штрафов) для работников, в связи с чем штат постоянно обновляется ("текучка кадров").
2. Порядка 15% (3553) работодателей указали в качестве сферы своей деятельности (как минимум одной) разработку программного обеспечения. Однако около 36% (8419) работодателей вообще не указали сферу своей деятельности. В то же время ряд компаний, например, 101 Интернет, 21vek.by, 2ГИС указали несколько сфер деятельности. Вероятно, компании, предоставляющей базу вакансий, имеет смысл в качестве обязательного ввести пункт описания сферы деятельности компании при регистрации на портале. Это важная информация не только для проведения статистических исследований, но и для фильтрации данных при осуществлении поиска вакансий соискателями.
3. В базе вакансий имеются работодатели без вакансий. Наибольшее количество таких работодателей (400), представленных в данной базе, - в России в целом. Логично, что на меньших (по площади и численности населения) территориях таких работодателей (как, скорее всего, и в целом работодателей) меньше. Наличие таких работодателей без вакансий в момент выгрузки данных из базы связано, скорее всего, с тем, что когда-то вакансии у этих работодателей в базе были, но закрылись, а данные о работодателе остались в базе.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [30]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [31]:
# текст запроса
query_3_1 = f'''
select count(id)
from public.vacancies
where Lower(name) LIKE '%data%' OR Lower(name) LIKE '%данн%'
'''

In [32]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [15]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [16]:
# текст запроса
query_3_1 = f'''
select count(id) 
from public.vacancies
where
((name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
OR ((name LIKE '%ML%')
AND (name NOT LIKE '%HTML%'))
OR (name ILIKE '%machine learning%')
OR (name ILIKE '%машинн%обучен%'))
AND (name ILIKE '%junior%' OR experience ILIKE '%Нет опыта%' OR employment ILIKE '%Стажировка%')

'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_13156\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [29]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [30]:
# текст запроса
query_3_1 = f'''
select count(id) 
from public.vacancies
where
((name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
OR ((name LIKE '%ML%')
AND (name NOT LIKE '%HTML%'))
OR (name ILIKE '%machine learning%')
OR (name ILIKE '%машинн%обучен%'))
AND (key_skills ILIKE '%SQL%' OR key_skills ILIKE 'postgres')

'''

In [31]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_13156\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [18]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [19]:
# текст запроса
query_3_1 = f'''
select count(id) 
from public.vacancies
where
((name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
OR ((name LIKE '%ML%')
AND (name NOT LIKE '%HTML%'))
OR (name ILIKE '%machine learning%')
OR (name ILIKE '%машинн%обучен%'))
AND (key_skills ILIKE '%python%')

'''

In [20]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_13156\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [38]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [39]:
query_3_1 = f'''
SELECT ROUND(AVG(1+length(key_skills) - length(REPLACE(key_skills, CHR(9), '' )))::numeric, 2) 
FROM public.vacancies  
where ((name ILIKE '%data scientist%') 
OR (name ILIKE '%data science%') 
OR (name ILIKE '%исследователь данных%') 
OR ((name LIKE '%ML%') AND (name NOT ILIKE '%HTML%')) 
OR (name ILIKE '%machine learning%') 
OR (name ILIKE '%машинн%обучен%') 
AND (key_skills IS NOT NULL))

'''


In [40]:

df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_13156\934135226.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [4]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [5]:
# текст запроса
query_3_1 = f'''
SELECT experience,
           Round(Avg(
                    (Coalesce(salary_from, salary_to) + Coalesce(salary_to, salary_from)) / 2
                )) AS ds_avg_salary
    FROM public.vacancies
    WHERE  (
            salary_to IS NOT NULL
            OR salary_from IS NOT NULL
           ) AND  
    ((name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name LIKE '%ML%')
    AND (name NOT LIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%'))

    GROUP BY experience

'''

In [6]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_9948\135049063.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


experience  ds_avg_salary
0           Нет опыта        74643.0
1  От 1 года до 3 лет       139675.0
2       От 3 до 6 лет       243115.0

***

# выводы по предметному анализу
Вакансии в области данных составляют около 3.6% от общего числа вакансий. Junior-специалисты в данной области не очень востребованы: на долю вакансий для джуниоров приходится всего 2,9% от всех вакансий для датасайентистов. Средние зарплаты в сфере данных варьируют от 74643.0 до 243115.0 руб. Их размер коррелирует с опытом работы.

# Общий вывод по проекту

Всего в базе hh на момент выгрузки данных было 49197 вакансий. Основными "поставщиками" вакансий являются такие крупные компании как Яндекс, Ростелеком, Тинькофф, СБЕР, Газпром нефть. Больше всего вакансий представлено в IT-сфере (запрос представлен ниже) и в крупных городах. 
Доля вакансий, имеющих отношение к данным, составила менее 4% от всех вакансий. Непосредственно вакансия с названием "Data Scientist" по частоте встречаемости (55) находится на 68 месте в ТОП-100. При этом средние значения заработной платы в сфере данных выше (варьируют от 74643 до 243115 руб.), чем средние значения таковых по всем вакансиям, представленным в базе (от 34699 до 197819 руб.) (См. выше - дополнительный анализ приведен в юните 4). В части зарплатных ожиданий это позволяет считать сферу деятельности, связанную с данными, достаточно привлекательной.  Однако при этом следует учитывать, что востребованными являются в основном специалисты, имеющие опыт работы. 
Учитывая рост объема данных в различных сферах, можно прогнозировать в будущем рост количества вакансий, связанных с данными. Однако есть вероятность, что данные профессии, став рядовыми, могут стать и менее оплачиваемыми. 

Наиболее часто встречающиеся вакансии

In [17]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [18]:
query_3_1 = f'''
SELECT DISTINCT name AS "Название вакансии", Count(*) AS "Частота встречаемости (количество)"
    FROM vacancies
    GROUP BY 1
    ORDER BY 2 DESC
    LIMIT 100

'''

In [19]:
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df

C:\Users\Павел\AppData\Local\Temp\ipykernel_9948\2285424339.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Название вакансии  \
0                             Системный администратор   
1                                      Программист 1С   
2                                     Бизнес-аналитик   
3                    Специалист технической поддержки   
4                                  Системный аналитик   
..                                                ...   
95                             Старший программист 1С   
96  Консультант онлайн-поддержки с техническими зн...   
97                                 Программист Python   
98                    Старший системный администратор   
99                                     Java developer   

    Частота встречаемости (количество  
0                                1761  
1                                1153  
2                                 488  
3                                 457  
4                                 432  
..                                ...  
95                                 42  
96                                 40  
97                                 40  
98                                 40  
99                                 39  

[100 rows x 2 columns]